tests/unit/amo/components/TestRatingsByStar.js

```diff
@@ -0,0 +1,289 @@
+    expect(counts.at(3)).toHaveText('2');
+    expect(counts.at(4)).toHaveText('1');
+  });
+
+  it('renders star counts', () => {
+    const grouping = {
+      5: 964,
+      4: 821,
+      3: 543,
+      2: 22,
+      1: 0,
+    };
+    const addon = addonForGrouping(grouping);
+    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const root = render({ addon });
+
+    const counts = root.find('.RatingsByStar-count');
+    expect(counts.at(0)).toHaveText(grouping[5].toString());
+    expect(counts.at(1)).toHaveText(grouping[4].toString());
+    expect(counts.at(2)).toHaveText(grouping[3].toString());
+    expect(counts.at(3)).toHaveText(grouping[2].toString());
+    expect(counts.at(4)).toHaveText(grouping[1].toString());
+  });
+
+  it('renders bar value widths based on total ratings', () => {
+    // Set up an add-on with one 5-star rating and one 4-star rating.
+    const grouping = {
+      5: 1,
+      4: 1,
+      3: 0,
+      2: 0,
+      1: 0,
+    };
+    const addon = addonForGrouping(grouping);
+    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const root = render({ addon });
+
+    const barValues = root.find('.RatingsByStar-barValue');
+    expect(barValues.at(0)).toHaveProp('style', { width: '50%' });
+    expect(barValues.at(1)).toHaveProp('style', { width: '50%' });
+    expect(barValues.at(2)).toHaveProp('style', { width: '0%' });
+    expect(barValues.at(3)).toHaveProp('style', { width: '0%' });
+    expect(barValues.at(4)).toHaveProp('style', { width: '0%' });
+  });
+
+  it('handles 100% bar values differently', () => {
+    // Set up an add-on with one 5-star rating.
+    const grouping = {
+      5: 1,
+      4: 0,
+      3: 0,
+      2: 0,
+      1: 0,
+    };
+    const addon = addonForGrouping(grouping);
+    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const root = render({ addon });
+
+    const barValues = root.find('.RatingsByStar-barValue');
+
+    expect(barValues.at(0)).toHaveProp('style', { width: '100%' });
+    expect(barValues.at(0)).not.toHaveClassName(
+      'RatingsByStar-barValueLessThan100',
+    );
+
+    expect(barValues.at(1)).toHaveProp('style', { width: '0%' });
+    expect(barValues.at(1)).toHaveClassName(
+      'RatingsByStar-barValueLessThan100',
+    );
+  });
+
+  it('renders 0% bar values when ratings do not exist', () => {
+    const grouping = { 5: 0, 4: 0, 3: 0, 2: 0, 1: 0 };
+    const addon = addonForGrouping(grouping, {
+      // This add-on does not have any ratings yet.
+      ratings: undefined,
+    });
+    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const root = render({ addon });
+
+    const barValues = root.find('.RatingsByStar-barValue');
+    for (const index of [0, 1, 2, 3, 4]) {
+      expect(barValues.at(index)).toHaveProp('style', { width: '0%' });
+    }
+  });
+
+  it('renders 0% bar values when ratings are all 0', () => {
```

Comment:
doesn't this mean ratings don't exist? (same as above?)

Reply:
One case is `addon.ratings.count = 0` and the other is `addon.ratings = undefined`. It's a little weird but I think we could legitimately encounter either case. The `addon.ratings` property is computed so there could be a split second when they are out of sync.

Reply:
I see..